File: lib/services/depth_anything_processor.dart
import 'dart:math' as math;
import 'dart:typed_data';
import 'dart:ui';

import 'package:image/image.dart' as img;
import 'package:tflite_flutter/tflite_flutter.dart';

class DepthAnalysis {
  const DepthAnalysis({
    required this.imageWidth,
    required this.imageHeight,
    required this.depth,
  });

  final int imageWidth;
  final int imageHeight;
  final DepthAnythingResult? depth;
}

class DepthAnythingResult {
  DepthAnythingResult({
    required this.width,
    required this.height,
    required this.values,
    required this.minValue,
    required this.maxValue,
  });

  final int width;
  final int height;
  final Float32List values;
  final double minValue;
  final double maxValue;

  double? metersForBox(
    Rect rect,
    int viewWidth,
    int viewHeight, {
    int stride = 4,
  }) {
    if (width <= 0 || height <= 0 || viewWidth <= 0 || viewHeight <= 0) {
      return null;
    }

    final normalizedLeft = _clamp(rect.left / viewWidth, 0.0, 1.0);
    final normalizedTop = _clamp(rect.top / viewHeight, 0.0, 1.0);
    final normalizedRight = _clamp(rect.right / viewWidth, 0.0, 1.0);
    final normalizedBottom = _clamp(rect.bottom / viewHeight, 0.0, 1.0);

    if (normalizedLeft >= normalizedRight || normalizedTop >= normalizedBottom) {
      return null;
    }

    final startX = (normalizedLeft * width).floor().clamp(0, width - 1);
    final endX = (normalizedRight * width).floor().clamp(startX, width - 1);
    final startY = (normalizedTop * height).floor().clamp(0, height - 1);
    final endY = (normalizedBottom * height).floor().clamp(startY, height - 1);
    final step = math.max(1, stride);

    final samples = <double>[];
    for (var y = startY; y <= endY; y += step) {
      final rowOffset = y * width;
      for (var x = startX; x <= endX; x += step) {
        final value = values[rowOffset + x];
        if (value.isFinite) {
          samples.add(value.toDouble());
        }
      }
    }

    if (samples.isEmpty) {
      return null;
    }

    samples.sort();
    final median = samples[samples.length ~/ 2];
    return _approximateMeters(median);
  }

  double _approximateMeters(double depthValue) {
    final clamped = _clamp(depthValue, minValue, maxValue);
    final normalized = (maxValue > minValue)
        ? (clamped - minValue) / (maxValue - minValue)
        : 0.5;
    final inverted = 1.0 - _clamp(normalized, 0.0, 1.0);
    const minDistance = 0.5;
    const maxDistance = 5.0;
    return minDistance + inverted * (maxDistance - minDistance);
  }

  double _clamp(double value, double min, double max) {
    if (value < min) return min;
    if (value > max) return max;
    return value;
  }
}

class DepthAnythingProcessor {
  DepthAnythingProcessor._(
    this._interpreter,
    this._inputShape,
    this._outputShape,
  )   : _inputHeight = _inputShape.length > 1 ? _inputShape[1] : 0,
        _inputWidth = _inputShape.length > 2 ? _inputShape[2] : 0,
        _inputChannels = _inputShape.length > 3 ? _inputShape[3] : 0,
        _outputHeight = _outputShape.length > 1 ? _outputShape[_outputShape.length - 3] : 0,
        _outputWidth = _outputShape.length > 2 ? _outputShape[_outputShape.length - 2] : 0,
        _outputChannels = _outputShape.length > 3 ? _outputShape[_outputShape.length - 1] : 1;

  final Interpreter _interpreter;
  final List<int> _inputShape;
  final List<int> _outputShape;
  final int _inputHeight;
  final int _inputWidth;
  final int _inputChannels;
  final int _outputHeight;
  final int _outputWidth;
  final int _outputChannels;

  static Future<DepthAnythingProcessor> create() async {
    final interpreter = await Interpreter.fromAsset('models/depth_anything.tflite');
    return DepthAnythingProcessor._(
      interpreter,
      interpreter.getInputTensor(0).shape,
      interpreter.getOutputTensor(0).shape,
    );
  }

  Future<DepthAnalysis?> estimate({
    Uint8List? imageBytes,
    img.Image? decodedImage,
  }) async {
    final img.Image? image = decodedImage ??
        (imageBytes != null ? img.decodeImage(imageBytes) : null);
    if (image == null || image.width <= 0 || image.height <= 0) {
      return null;
    }

    final depth = _runInference(image);
    return DepthAnalysis(
      imageWidth: image.width,
      imageHeight: image.height,
      depth: depth,
    );
  }

  DepthAnythingResult? _runInference(img.Image image) {
    if (_inputHeight <= 0 || _inputWidth <= 0 || _inputChannels <= 0) {
      return null;
    }

    final resized = img.copyResize(
      image,
      width: _inputWidth,
      height: _inputHeight,
      interpolation: img.Interpolation.linear,
    );

    final input = List.generate(
      1,
      (_) => List.generate(
        _inputHeight,
        (_) => List.generate(
          _inputWidth,
          (_) => List<double>.filled(_inputChannels, 0.0),
        ),
      ),
    );

    for (var y = 0; y < _inputHeight; y++) {
      for (var x = 0; x < _inputWidth; x++) {
        final pixel = resized.getPixel(x, y);
        final channels = input[0][y][x];
        if (_inputChannels >= 1) {
          channels[0] = pixel.rNormalized.toDouble();
        }
        if (_inputChannels >= 2) {
          channels[1] = pixel.gNormalized.toDouble();
        }
        if (_inputChannels >= 3) {
          channels[2] = pixel.bNormalized.toDouble();
        }
      }
    }

    final depthData = Float32List(_outputWidth * _outputHeight);
    double minValue = double.infinity;
    double maxValue = double.negativeInfinity;

    if (_outputShape.length == 3) {
      final output = List.generate(
        1,
        (_) => List.generate(
          _outputHeight,
          (_) => List<double>.filled(_outputWidth, 0.0),
        ),
      );
      _interpreter.run(input, output);
      var index = 0;
      for (var y = 0; y < _outputHeight; y++) {
        final row = output[0][y];
        for (var x = 0; x < _outputWidth; x++) {
          final value = row[x];
          final floatValue = value.toDouble();
          depthData[index++] = floatValue;
          if (floatValue.isFinite) {
            if (floatValue < minValue) minValue = floatValue;
            if (floatValue > maxValue) maxValue = floatValue;
          }
        }
      }
    } else {
      final channels = math.max(1, _outputChannels);
      final output = List.generate(
        1,
        (_) => List.generate(
          _outputHeight,
          (_) => List.generate(
            _outputWidth,
            (_) => List<double>.filled(channels, 0.0),
          ),
        ),
      );
      _interpreter.run(input, output);
      var index = 0;
      for (var y = 0; y < _outputHeight; y++) {
        final row = output[0][y];
        for (var x = 0; x < _outputWidth; x++) {
          final channelValues = row[x];
          final value = channelValues[0];
          final floatValue = value.toDouble();
          depthData[index++] = floatValue;
          if (floatValue.isFinite) {
            if (floatValue < minValue) minValue = floatValue;
            if (floatValue > maxValue) maxValue = floatValue;
          }
        }
      }
    }

    if (!minValue.isFinite || !maxValue.isFinite) {
      minValue = 0.0;
      maxValue = 1.0;
    }

    return DepthAnythingResult(
      width: _outputWidth,
      height: _outputHeight,
      values: depthData,
      minValue: minValue,
      maxValue: maxValue,
    );
  }

  void close() {
    _interpreter.close();
  }
}

File: android/app/src/main/kotlin/com/ultralytics/yolo_example/DepthAnythingProcessor.kt
package com.ultralytics.yolo_example

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.media.Image
import android.util.Log
import com.google.ar.core.Frame
import com.google.ar.core.exceptions.NotYetAvailableException
import io.flutter.FlutterInjector
import java.io.ByteArrayOutputStream
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel
import kotlin.math.max
import kotlin.math.min
import org.tensorflow.lite.Interpreter

class DepthAnythingProcessor(private val context: Context) {
    companion object {
        private const val TAG = "DepthAnythingProcessor"
        private const val MODEL_ASSET_PATH = "assets/models/depth_anything.tflite"
    }

    @Volatile
    private var interpreter: Interpreter? = null
    @Volatile
    private var inputShape: IntArray = intArrayOf()
    @Volatile
    private var outputShape: IntArray = intArrayOf()
    @Volatile
    private var cachedTimestamp: Long = -1L
    @Volatile
    private var cachedResult: DepthAnythingResult? = null

    fun estimate(frame: Frame): DepthAnythingResult? {
        val timestamp = frame.timestamp
        val existing = cachedResult
        if (timestamp != 0L && timestamp == cachedTimestamp && existing != null) {
            return existing
        }

        val image = try {
            frame.acquireCameraImage()
        } catch (_: NotYetAvailableException) {
            return existing
        } catch (error: Exception) {
            Log.w(TAG, "Failed to acquire camera image: ${error.localizedMessage}")
            return existing
        }

        val result = try {
            estimate(image)
        } finally {
            image.close()
        }

        if (result != null && timestamp != 0L) {
            cachedTimestamp = timestamp
            cachedResult = result
        }
        return result
    }

    fun clear() {
        cachedTimestamp = -1L
        cachedResult = null
    }

    private fun estimate(image: Image): DepthAnythingResult? {
        val interpreter = ensureInterpreter() ?: return null
        val bitmap = image.toBitmap() ?: return null

        val inputHeight = inputShape.getOrNull(1) ?: return null
        val inputWidth = inputShape.getOrNull(2) ?: return null

        val scaled = Bitmap.createScaledBitmap(bitmap, inputWidth, inputHeight, true)
        val inputBuffer = ByteBuffer.allocateDirect(4 * inputWidth * inputHeight * 3)
            .order(ByteOrder.nativeOrder())
        for (y in 0 until inputHeight) {
            for (x in 0 until inputWidth) {
                val pixel = scaled.getPixel(x, y)
                val r = ((pixel shr 16) and 0xFF) / 255f
                val g = ((pixel shr 8) and 0xFF) / 255f
                val b = (pixel and 0xFF) / 255f
                inputBuffer.putFloat(r)
                inputBuffer.putFloat(g)
                inputBuffer.putFloat(b)
            }
        }
        inputBuffer.rewind()

        val outputHeight = outputShape.getOrNull(1) ?: return null
        val outputWidth = outputShape.getOrNull(2) ?: return null
        val outputBuffer = ByteBuffer.allocateDirect(4 * outputWidth * outputHeight)
            .order(ByteOrder.nativeOrder())
        interpreter.run(inputBuffer, outputBuffer)
        outputBuffer.rewind()
        val depthValues = FloatArray(outputWidth * outputHeight)
        outputBuffer.asFloatBuffer().get(depthValues)

        var minValue = Float.POSITIVE_INFINITY
        var maxValue = Float.NEGATIVE_INFINITY
        for (value in depthValues) {
            if (!value.isFinite()) continue
            if (value < minValue) minValue = value
            if (value > maxValue) maxValue = value
        }
        if (minValue == Float.POSITIVE_INFINITY || maxValue == Float.NEGATIVE_INFINITY) {
            minValue = 0f
            maxValue = 1f
        }

        return DepthAnythingResult(outputWidth, outputHeight, depthValues, minValue, maxValue)
    }

    private fun ensureInterpreter(): Interpreter? {
        var interpreter = this.interpreter
        if (interpreter != null) return interpreter
        synchronized(this) {
            interpreter = this.interpreter
            if (interpreter != null) return interpreter
            return try {
                val loader = FlutterInjector.instance().flutterLoader()
                val assetKey = loader.getLookupKeyForAsset(MODEL_ASSET_PATH)
                val assetManager = context.assets
                assetManager.openFd(assetKey).use { descriptor ->
                    FileInputStream(descriptor.fileDescriptor).use { inputStream ->
                        val startOffset = descriptor.startOffset
                        val declaredLength = descriptor.declaredLength
                        val channel = inputStream.channel
                        val mapped = channel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
                        val options = Interpreter.Options().apply { setNumThreads(2) }
                        val newInterpreter = Interpreter(mapped, options)
                        inputShape = newInterpreter.getInputTensor(0).shape()
                        outputShape = newInterpreter.getOutputTensor(0).shape()
                        this.interpreter = newInterpreter
                        newInterpreter
                    }
                }
            } catch (error: Exception) {
                Log.e(TAG, "Failed to load depth model", error)
                null
            }
        }
    }

    private fun Image.toBitmap(): Bitmap? {
        return try {
            if (format != ImageFormat.YUV_420_888) return null
            val yBuffer = planes[0].buffer
            val uBuffer = planes[1].buffer
            val vBuffer = planes[2].buffer
            val ySize = yBuffer.remaining()
            val uSize = uBuffer.remaining()
            val vSize = vBuffer.remaining()
            val nv21 = ByteArray(ySize + uSize + vSize)
            yBuffer.get(nv21, 0, ySize)
            vBuffer.get(nv21, ySize, vSize)
            uBuffer.get(nv21, ySize + vSize, uSize)
            val yuvImage = YuvImage(nv21, ImageFormat.NV21, width, height, null)
            val out = ByteArrayOutputStream()
            yuvImage.compressToJpeg(Rect(0, 0, width, height), 80, out)
            val bytes = out.toByteArray()
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        } catch (error: Exception) {
            Log.w(TAG, "Failed to convert image to bitmap: ${error.localizedMessage}")
            null
        }
    }
}

data class DepthAnythingResult(
    val width: Int,
    val height: Int,
    val values: FloatArray,
    val minValue: Float,
    val maxValue: Float,
) {
    fun metersForBox(det: Det, viewWidth: Int, viewHeight: Int, stride: Int = 4): Float? {
        if (width <= 0 || height <= 0 || viewWidth <= 0 || viewHeight <= 0) return null
        val normalizedLeft = max(0f, det.boxViewPx.left / viewWidth.toFloat())
        val normalizedTop = max(0f, det.boxViewPx.top / viewHeight.toFloat())
        val normalizedRight = min(1f, det.boxViewPx.right / viewWidth.toFloat())
        val normalizedBottom = min(1f, det.boxViewPx.bottom / viewHeight.toFloat())
        if (normalizedLeft >= normalizedRight || normalizedTop >= normalizedBottom) return null

        val samples = mutableListOf<Float>()
        val startX = (normalizedLeft * width).toInt().coerceIn(0, width - 1)
        val endX = (normalizedRight * width).toInt().coerceIn(startX, width - 1)
        val startY = (normalizedTop * height).toInt().coerceIn(0, height - 1)
        val endY = (normalizedBottom * height).toInt().coerceIn(startY, height - 1)
        val step = max(1, stride)
        for (y in startY..endY step step) {
            for (x in startX..endX step step) {
                val value = values[y * width + x]
                if (value.isFinite()) {
                    samples.add(value)
                }
            }
        }
        if (samples.isEmpty()) return null
        samples.sort()
        val median = samples[samples.size / 2]
        return approximateMeters(median)
    }

    private fun approximateMeters(depthValue: Float): Float {
        val clamped = depthValue.coerceIn(minValue, maxValue)
        val normalized = if (maxValue > minValue) {
            (clamped - minValue) / (maxValue - minValue)
        } else {
            0.5f
        }
        val inverted = 1f - normalized.coerceIn(0f, 1f)
        return MIN_DISTANCE + inverted * (MAX_DISTANCE - MIN_DISTANCE)
    }

    companion object {
        private const val MIN_DISTANCE = 0.5f
        private const val MAX_DISTANCE = 5.0f
    }
}
